src/Language/PureScript/Linter/Exhaustive.hs

```diff
@@ -0,0 +1,228 @@
+-----------------------------------------------------------------------------
+--
+-- Module      :  Language.PureScript.Exhaustive
+-- Copyright   :  (c) 2013-14 Phil Freeman, (c) 2014 Gary Burgess, and other contributors
+-- License     :  MIT
+--
+-- Maintainer  :  Phil Freeman <paf31@cantab.net>
+-- Stability   :  experimental
+-- Portability :
+--
+-----------------------------------------------------------------------------
+
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Language.PureScript.Linter.Exhaustive 
+  ( checkExhaustive
+  , checkExhaustiveModule
+  ) where
+
+import qualified Data.Map as M
+import Data.Maybe (fromMaybe)
+import Data.List (foldl', sortBy)
+import Data.Function (on)
+
+import Control.Applicative
+import Control.Monad.Writer.Class
+
+import Language.PureScript.AST.Binders
+import Language.PureScript.AST.Declarations
+import Language.PureScript.Environment
+import Language.PureScript.Names as P
+import Language.PureScript.Kinds
+import Language.PureScript.Types as P
+import Language.PureScript.Errors
+
+import Language.PureScript.AST.Traversals (everywhereOnValuesTopDownM)
+
+-- |
+-- Qualifies a propername from a given qualified propername
+--
+qualifyProperName :: ProperName -> Qualified ProperName -> Qualified ProperName
+qualifyProperName pn qpn = Qualified (Just mn) pn
+  where
+  (mn, _) = qualify defmn qpn
+  defmn = moduleNameFromString "Main"
+
+-- |
+-- Given an environment and a datatype or newtype name,
+-- this function returns the associated data constructors if it's the case of a datatype
+-- where: - ProperName is the name of the constructor (for example, "Nothing" in Maybe)
+--        - [Type] is the list of arguments, if it has (for example, "Just" has [TypeVar "a"])
+--
+getConstructors :: Environment -> (Qualified ProperName) -> [(ProperName, [Type])]
+getConstructors env n = extractConstructors lnte
+  where
+  qpn :: Qualified ProperName
+  qpn = getConsDataName n
+
+  getConsDataName :: (Qualified ProperName) -> (Qualified ProperName)
+  getConsDataName con = qualifyProperName nm con
+    where
+    nm = case getConsInfo con of
+           Nothing -> error $ "ProperName " ++ show con ++ " not in the scope of the current environment in getConsDataName."
+           Just (_, pm, _, _) -> pm
+
+  getConsInfo :: (Qualified ProperName) -> Maybe (DataDeclType, ProperName, Type, [Ident])
+  getConsInfo con = M.lookup con dce
+    where
+    dce :: M.Map (Qualified ProperName) (DataDeclType, ProperName, Type, [Ident])
+    dce = dataConstructors env
+
+  lnte :: Maybe (Kind, TypeKind) 
+  lnte = M.lookup qpn (types env)
+
+  extractConstructors :: Maybe (Kind, TypeKind) -> [(ProperName, [Type])]
+  extractConstructors (Just (_, DataType _ pt)) = pt
+  extractConstructors _ = error ""
+
+-- |
+-- Replicates a wildcard binder
+--
+initialize :: Int -> [Binder]
+initialize l = replicate l NullBinder
+
+-- |
+--
+--
+completeMissing :: Ord a =>
```

Comment:
Maybe this could be called `genericMerge` or something like that? Also, could you please comment it?